execution.bs

```diff
@@ -3173,67 +3196,91 @@ Insert this section as a new subclause between Header `<stop_token>` synopsis <b
+    `CB` that is callable with no arguments, the type `T::callback_type<CB>` shall
+    be well-formed and name the stop callback type to use to register a callback
+    to be executed if a stop request is ever made on a `stoppable_token` of type
+    `T`. The `stoppable_token_for` concept checks for a stop token type compatible
+    with a given callback type. The `unstoppable_token` concept checks for a stop
+    token type that does not allow stopping.
 
 <pre>
-template&lt;typename T>
+template&lt;class T>
   concept stoppable_token =
-    copy_constructible&lt;T> &&
-    move_constructible&lt;T> &&
-    is_nothrow_copy_constructible_v&lt;T> &&
-    is_nothrow_move_constructible_v&lt;T> &&
+    copyable&lt;T> &&
     equality_comparable&lt;T> &&
-    requires (const T& token) {
-      { token.stop_requested() } noexcept -> boolean-testable;
-      { token.stop_possible() } noexcept -> boolean-testable;
+    requires (const T t) {
+      { T{t} } noexcept;
+      { t.stop_requested() } noexcept -> <i>boolean-testable</i>;
+      { t.stop_possible() } noexcept -> <i>boolean-testable</i>;
       typename <i>check-type-alias-exists</i>&lt;T::template callback_type>;
     };
 
-template&lt;typename T, typename CB, typename Initializer = CB>
+template&lt;class T, class CB, class Initializer = CB>
   concept stoppable_token_for =
     stoppable_token&lt;T> &&
     invocable&lt;CB> &&
-    requires {
-      typename T::template callback_type&lt;CB>;
-    } &&
     constructible_from&lt;CB, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, T, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, T&, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, const T, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, const T&, Initializer>;
+    requires (const T t) {
+      typename stop_callback_for_t&lt;T, CB>;
+      []&lt;class T>(T&& t)
+        requires constructible_from&lt;stop_callback_for_t&lt;T, CB>, U, Initializer>
+      {}(t);
+    };
 
-template&lt;typename T>
+template&lt;class T>
   concept unstoppable_token =
     stoppable_token&lt;T> &&
     requires {
-      { T::stop_possible() } -> boolean-testable;
-    } &&
-    (!T::stop_possible());
+      { bool_constant&lt;T::stop_possible()>{} } -> same_as&lt;false_type>;
+    };
 </pre>
 
-2. Let `t` and `u` be distinct object of type `T`. The type `T` models `stoppable_token` only if:
+<div class="ed-note">LWG directed me to replace `T::stop_possible()` with `t.stop_possible()` because
+of the recent `constexpr` changes in [[P2280r2|P2280R2]]. However, even with those changes, a nested
+requirement like `requires (!t.stop_possible())`, where `t` is an argument in the requirement-parameter-list, is ill-formed according to
+<a href="http://eel.is/c++draft/expr.prim.req#nested-2.sentence-1">[expr.prim.req.nested/p2]</a>:
```

Comment:
I don't think that testing `t.stop_possible()` is the right thing to do here.

Let's say that `in_place_stop_token` were constexpr default-constructible to the invalid stop-token.
This would then potentially mean that `in_place_stop_token{}.stop_possible()` could be constexpr and would return `false` despite there being other instances of `in_place_stop_token` for which `stop_possible()` might return `true`.

Reply:
If `t.stop_possible()` is conatexpr with unknown (run-time) t, it means that it does not depent on any stare in t. Something like size() of array.